### PR TITLE
Option to force pthread_create to succeed

### PR DESCRIPTION
**Pull request by hernanponcedeleon**
Some programs rely on some sort of fairness where it is expected that every thread is eventually executed.
Calling `pthread_create` can fail and we model this faithfully, but it can then violate the assumption above.
This PR prints a warning when there is a violation and some of the threads did not get executed (i.e. the jump coming from the compilation of `Start` was taken).
It also adds an option to force `pthread_create` to succeed (by assuming that the register of the load coming from a `Start` event is not zero).

### CODE REVIEW

**Comment by ThomasHaas** [comment]
The forcing of thread starts via assumes has some problems you need to be aware of.
If the main thread (or any other thread) does conditionally start a thread (i.e., `pthread_create` is behind some conditionals), then you can get unexpected results, cause the `assume` will remove executions that did not even spawn the thread at all.
Technically, a thread spuriously fails if despite `pthread_create` being called, the thread does not start.

I'm not sure if this is possible, but you could try to use `ExecutionStatus` to track within a thread if the corresponding `pthread_create` of the main thread (or whoever created the thread) got executed, and then essentially assume `pthread_create.exec => Loaded value != 0`.

**Comment by hernanponcedeleon** [comment]
The only wrong result should be fixed by #314. @ThomasHaas if you are ok with the rest of the code, I will merge once the CI passes

**Comment by ThomasHaas** [comment]
I haven't yet looked into the details of this PR. Let me check the PR more in detail before we merge.

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/boogie/PthreadPool.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/boogie/PthreadPool.java:22]: "This is needed during the compilation pass to match a Start with its corresponding Create. Both events access the same address, i.e. the communication channel (cc), which we only use for modeling purposes. At this point the Create was already compiled and thus we use an annotation event which remains after the compilation." ... Would this be enough? I want to avoid long variable names.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java:3]: We could fully remove it and use simply a skip. That might be a better option than making this more general when we don't have another use case.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/lang/pthread/Start.java
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/lang/pthread/Start.java:16]: If by "the store would not be needed" you mean we could remove it (or not having it in the first place), I don't see how you could use an RF-edge

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/lang/pthread/Start.java:16]: I should have formulated it better. I'm saying that the whole logic of starting a thread does not actually need the rf-edge in principle, and in particular, the communicated value doesn't matter either.
However, we still need the rf-edge to get proper propagation behavior between the threads.
But one could imagine to use e.g. a co-edge or po-edge to get the right propagation semantics.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java:3]: I would want to have something descriptive. Maybe just a very generic `StringAnnotation` event that is essentially a `skip` but with associated text.
Then you would use `EventFactory.newStringAnnotation("Spawning thread " + tId/name)` which would be printed like this:
```
===== Spawning thread X =====
Fence;
store(...)
Fence;

```

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java:3]: Ok. But instead of creating yet another class which will simply extend `Skip`, with not add a label argument to `Skip`, return the label when calling `toString()` and in the factory allow `newSkip(label)` and `newSkip()` which will simply return `new Skip("skip")`

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java:3]: Well, the event would still extend `CodeAnnotation` and hence inherit the `Annotation` tag.
I think it makes sense to distinguish between plain skip events and code annotations.

Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java:3]: Ok, but does it make sense to keep code annotations abstract and keep adding events that extend it? So far only FunCall and FunRet extend it and both have a string, so we can simply make the string part of the CodeAnnotation and use this event for the matcher (and simplify the other two classes). Then CodeAnnotation would have the tag and the string annotation we want to use in the code, and FunRet and FunCall just a minimal logic about the function name which is used in the simplification pass

Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/annotations/MatcherCreateStart.java:3]: I would keep `CodeAnnotation` abstract because not every annotation needs a string (we also have a set of loop annotations none of which take strings as input).
But it might make sense to create a single subclass that does nothing but hold a string (`StringAnnotation` or `Comment`?). We might then use this class whenever we just want to place a generic marker or inherit from it for specialized string-based markers (like `FunCall` and `FunRet`).

**Review by ThomasHaas — APPROVED** [review]
LGTM

**Pull request merged**
